sources/imdsclient/src/lib.rs

```diff
@@ -221,92 +241,114 @@ impl ImdsClient {
-                StatusCode::NOT_FOUND => return Ok(None),
+                    code => {
+                        let response_body = response
+                            .bytes()
+                            .await
+                            .context(error::ResponseBody {
+                                method: "GET",
+                                uri: &uri,
+                                code,
+                            })?
+                            .to_vec();
 
-                // IMDS returns 401 if the session token is expired or invalid
-                StatusCode::UNAUTHORIZED => {
-                    info!("Session token is invalid or expired");
-                    self.refresh_token().await?;
-                    info!("Refreshed session token");
-                    continue;
-                }
+                        let response_str = printable_string(&response_body);
 
-                StatusCode::REQUEST_TIMEOUT => {
-                    info!("Retrying request");
-                    continue;
-                }
+                        trace!("Response: {:?}", response_str);
 
-                code => {
-                    let response_body = response
-                        .bytes()
-                        .await
-                        .context(error::ResponseBody {
+                        error::Response {
                             method: "GET",
                             uri: &uri,
                             code,
-                        })?
-                        .to_vec();
-
-                    let response_str = printable_string(&response_body);
-
-                    trace!("Response: {:?}", response_str);
-
-                    return error::Response {
-                        method: "GET",
-                        uri: &uri,
-                        code,
-                        response_body: response_str,
+                            response_body: response_str,
+                        }
+                        .fail()
                     }
-                    .fail();
                 }
-            }
-        }
-    }
-
-    /// Fetches a new session token and adds it to the current ImdsClient.
-    async fn refresh_token(&mut self) -> Result<()> {
-        self.session_token = fetch_token(&self.client, &self.imds_base_uri).await?;
-        Ok(())
+            }),
+        )
+        .await
+        .context(error::TimeoutFetchIMDS)?
+    }
+
+    /// Fetches a new session token and writes it to the current ImdsClient.
+    async fn write_token(&self) -> Result<String> {
+        let session_token =
+            match fetch_token(&self.client, &self.imds_base_uri, &self.retry_timeout).await? {
+                Some(session_token) => {
+                    *self
+                        .session_token
+                        .write()
+                        .map_err(|_| error::Error::FailedWriteToken {})? =
+                        Some(session_token.clone());
```

Comment:
nit: The prefix `*` notation in Rust for dereferencing can be a bit difficult to parse for readers. In this case, the way that `RwLock` provides you write access is by returning an `RwLockWriteGuard<String>` which protects our value. `RwLockWriteGuard<T>` gives you access to the internal value by implementing `Deref`.

It might read more cleanly to call the `deref()` method over using the `*` prefix.


```suggestion
                    self
                        .session_token
                        .write()
                        .map_err(|_| error::Error::FailedWriteToken {})?
                        .deref() =
                        Some(session_token.clone());
```

Reply:
After talking with @jpculp, we concluded that this won't work! It's because I misunderstood how `Deref` works. See [this SA answer](https://stackoverflow.com/a/62506306):

> If x has type &T or &mut T, then *x has type T.
>
> What if x has some other type S? If S implements Deref or DerefMut, then Rust will call .deref() or .deref_mut() on x, and dereference that.